### PR TITLE
SQHELM-107 Add a configurable Prometheus PodMonitor resource

### PR DESCRIPTION
**Pull request by carminevassallo**
Co-authored-by: @dpunkturban

This PR is based on the contribution made by Dennis Urban in the context of [this PR](https://github.com/SonarSource/helm-chart-sonarqube/pull/273).

### CODE REVIEW

File: charts/sonarqube-dce/templates/prometheus-podmonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{{- if and .Values.ApplicationNodes.prometheusExporter.enabled .Values.ApplicationNodes.prometheusExporter.podMonitor.enabled }}
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: {{ template "sonarqube.name" . }}
+  namespace: {{ .Values.ApplicationNodes.prometheusExporter.podMonitor.namespace | quote }}
+  labels:
+    app: {{ template "sonarqube.name" . }}
+spec:
+   {{- if  .Values.ApplicationNodes.prometheusExporter.podMonitor.jobLabel }}
```
Comment (jCOTINEAU) [inline charts/sonarqube-dce/templates/prometheus-podmonitor.yaml:10]: extra space after if ?

File: charts/sonarqube-dce/templates/prometheus-podmonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    matchNames:
+    - {{ .Release.Namespace }}
+  selector:
+    matchLabels:
+      app: {{ template "sonarqube.name" . }}
+  podMetricsEndpoints:
+  - port: monitoring-ce
+    path: /
+    scheme: http
+    {{- if  .Values.ApplicationNodes.prometheusExporter.podMonitor.interval }}
```
Comment (jCOTINEAU) [inline charts/sonarqube-dce/templates/prometheus-podmonitor.yaml:23]: extra space after if

File: charts/sonarqube-dce/templates/prometheus-podmonitor.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  {{- end }}
+  namespaceSelector:
+    matchNames:
+    - {{ .Release.Namespace }}
+  selector:
+    matchLabels:
+      app: {{ template "sonarqube.name" . }}
+  podMetricsEndpoints:
+  - port: monitoring-ce
+    path: /
```
Comment (jCOTINEAU) [inline charts/sonarqube-dce/templates/prometheus-podmonitor.yaml:21]: Are we sure this '/' path will always be true ? 

Is there an SQ parameter that could modify that ? probably not that kind of stuff is handled by the reverse proxy i believe ?

File: charts/sonarqube-dce/templates/prometheus-podmonitor.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    {{- if  .Values.ApplicationNodes.prometheusExporter.podMonitor.interval }}
+    interval: {{ .Values.ApplicationNodes.prometheusExporter.podMonitor.interval }}
+    {{- end }}
+    {{- if .Values.ApplicationNodes.prometheusExporter.podMonitor.scrapeTimeout }}
+    scrapeTimeout: {{ .Values.ApplicationNodes.prometheusExporter.podMonitor.scrapeTimeout }}
+    {{- end }}
+  - port: monitoring-web
+    path: /
+    scheme: http
+    {{- if  .Values.ApplicationNodes.prometheusExporter.podMonitor.interval }}
```
Comment (jCOTINEAU) [inline charts/sonarqube-dce/templates/prometheus-podmonitor.yaml:32]: extra space after if

File: charts/sonarqube/templates/prometheus-podmonitor.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{{- if and .Values.prometheusExporter.enabled .Values.prometheusExporter.podMonitor.enabled }}
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: {{ template "sonarqube.name" . }}
+  namespace: {{ .Values.prometheusExporter.podMonitor.namespace | quote }}
+  labels:
+    app: {{ template "sonarqube.name" . }}
+spec:
+   {{- if  .Values.prometheusExporter.podMonitor.jobLabel }}
```
Comment (jCOTINEAU) [inline charts/sonarqube/templates/prometheus-podmonitor.yaml:10]: same extra if as in dce

File: charts/sonarqube-dce/templates/prometheus-podmonitor.yaml (hunk 3 above)
Comment (carminevassallo) [reply charts/sonarqube-dce/templates/prometheus-podmonitor.yaml:21]: I do not think there is a property that modifies it. But, while you were talking about this, I also realize that `monitoring-ce` and `monitoring-web` do not really expose prometheus metrics. Shall we keep just the real prometheus endpoint in the resource (i.e., `/api/monitoring/metrics`) as we specify in the [docs](https://docs.sonarqube.org/latest/instance-administration/monitoring/)?

Comment (jCOTINEAU) [reply charts/sonarqube-dce/templates/prometheus-podmonitor.yaml:21]: Yes there is something in the picture i still do not understand.

We have an init container that inject the prometheus exporter, that is no longer needed i believe, as it is now fully included in SQ no ? 

That's probably how it was done before, so yup i would agree we remove every mention to those things.

Comment (carminevassallo) [reply charts/sonarqube-dce/templates/prometheus-podmonitor.yaml]: It's indeed a tricky topic. As discussed, we both need to convert the JMX metrics and support the native prometheus metrics. For now, I was just decoupling the prometheus exporter and the prometheus PodMonitor. In my opinion, they are independent. Furthermore, I made sure that our podmonitor only supports the native prometheus endpoint. We need to discuss how to support such Prometheus exporter in the future...

File: charts/sonarqube/templates/secret.yaml [hunk 6]
```diff
@@ -29,6 +29,22 @@ data:
   SONAR_WEB_SYSTEMPASSCODE: {{ .Values.monitoringPasscode | b64enc | quote }}
 {{- end }}
 ---
+{{- if and .Values.monitoringPasscode .Values.prometheusMonitoring.podMonitor.enabled}}
+apiVersion: v1
```
Comment (jCOTINEAU) [inline charts/sonarqube/templates/secret.yaml:33]: Is that Secret a duplicate of the one above ? on line 18 to 30 ? Apart from the confition it looks like the same, and both would be applied when `if and .Values.monitoringPasscode .Values.prometheusMonitoring.podMonitor.enabled` condition is true ?

Comment (carminevassallo) [reply charts/sonarqube/templates/secret.yaml:33]: Unfortunately, we need the same secret both in the release and the podmonitor namespaces. But such duplicate code can be refactored. I've just defined [a task](https://sonarsource.atlassian.net/browse/SQHELM-117) for it.

**Review by jCOTINEAU — APPROVED** [review]
lgtm.

**Pull request merged**